File: 05-seguranca-ctf/backend/php/src/Controllers/OrderController.php
<?php

declare(strict_types=1);

namespace App\Controllers;

use App\Database;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class OrderController
{
    public function index(Request $request, Response $response): Response
    {
        $claims = $request->getAttribute('jwt_claims');
        $db = Database::getInstance();

        if ($claims['role'] === 'admin') {
            $sqlQuery = "SELECT * FROM orders ORDER BY id DESC";
            $orders = $db->query($sqlQuery)->fetchAll();
        } else {
            $sqlQuery = "SELECT * FROM orders WHERE user_id = {$claims['sub']} ORDER BY id DESC";
            $stmt = $db->prepare("SELECT * FROM orders WHERE user_id = ? ORDER BY id DESC");
            $stmt->execute([$claims['sub']]);
            $orders = $stmt->fetchAll();
        }

        Database::logRequest([
            'method' => 'GET', 'path' => '/api/orders',
            'queryParams' => '', 'body' => '',
            'statusCode' => 200, 'sqlQuery' => $sqlQuery,
            'responsePreview' => count($orders) . ' orders',
            'ip' => self::getIp($request),
        ]);

        $response->getBody()->write(json_encode([
            'orders' => $orders,
            'count'  => count($orders),
        ]));
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function show(Request $request, Response $response, array $args): Response
    {
        $id = $args['id'];
        $db = Database::getInstance();

        $sqlQuery = "SELECT o.*, u.nome as user_name, u.email as user_email FROM orders o JOIN users u ON o.user_id = u.id WHERE o.id = $id";

        $stmt = $db->prepare(
            "SELECT o.*, u.nome as user_name, u.email as user_email FROM orders o JOIN users u ON o.user_id = u.id WHERE o.id = ?"
        );
        $stmt->execute([$id]);
        $order = $stmt->fetch();

        Database::logRequest([
            'method' => 'GET', 'path' => "/api/orders/$id",
            'queryParams' => '', 'body' => '',
            'statusCode' => $order ? 200 : 404, 'sqlQuery' => $sqlQuery,
            'responsePreview' => $order ? "order {$order['id']}" : 'not found',
            'ip' => self::getIp($request),
        ]);

        if (!$order) {
            $response->getBody()->write(json_encode([
                'error' => 'Not Found',
                'message' => 'Pedido nao encontrado',
            ]));
            return $response->withStatus(404)->withHeader('Content-Type', 'application/json');
        }

        $response->getBody()->write(json_encode($order));
        return $response->withHeader('Content-Type', 'application/json');
    }

    public function create(Request $request, Response $response): Response
    {
        $claims = $request->getAttribute('jwt_claims');

        $body = $request->getParsedBody();
        if ($body === null) {
            $raw = (string) $request->getBody();
            $body = json_decode($raw, true);
        }
        if (!is_array($body)) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'JSON invalido',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $items = $body['items'] ?? [];
        $total = $body['total'] ?? 0;

        if (empty($items) || $total <= 0) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'Pedido deve conter itens e total valido',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $db = Database::getInstance();
        $stmt = $db->prepare("INSERT INTO orders (user_id, total, status) VALUES (?, ?, 'pendente')");
        $stmt->execute([$claims['sub'], $total]);

        $newOrderId = $db->lastInsertId();
        $stmt = $db->prepare("SELECT * FROM orders WHERE id = ?");
        $stmt->execute([$newOrderId]);
        $newOrder = $stmt->fetch();

        Database::logRequest([
            'method' => 'POST', 'path' => '/api/orders',
            'queryParams' => '', 'body' => json_encode($body),
            'statusCode' => 201,
            'sqlQuery' => "INSERT INTO orders (user_id, total, status) VALUES ({$claims['sub']}, $total, 'pendente')",
            'responsePreview' => "order $newOrderId created",
            'ip' => self::getIp($request),
        ]);

        $response->getBody()->write(json_encode([
            'message' => 'Pedido criado com sucesso',
            'order'   => $newOrder,
        ]));
        return $response->withStatus(201)->withHeader('Content-Type', 'application/json');
    }

    public function applyCoupon(Request $request, Response $response, array $args): Response
    {
        $id = $args['id'];

        $body = $request->getParsedBody();
        if ($body === null) {
            $raw = (string) $request->getBody();
            $body = json_decode($raw, true);
        }
        if (!is_array($body)) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'JSON invalido',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $code = $body['code'] ?? '';
        if (!$code) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'Codigo do cupom obrigatorio',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        $db = Database::getInstance();

        $stmt = $db->prepare("SELECT * FROM coupons WHERE code = ? AND active = 1");
        $stmt->execute([$code]);
        $coupon = $stmt->fetch();

        if (!$coupon) {
            $response->getBody()->write(json_encode([
                'error' => 'Not Found',
                'message' => 'Cupom nao encontrado',
            ]));
            return $response->withStatus(404)->withHeader('Content-Type', 'application/json');
        }

        if ($coupon['uses'] >= $coupon['max_uses']) {
            $response->getBody()->write(json_encode([
                'error' => 'Bad Request',
                'message' => 'Cupom esgotado',
            ]));
            return $response->withStatus(400)->withHeader('Content-Type', 'application/json');
        }

        usleep(100000);

        $stmt = $db->prepare("UPDATE coupons SET uses = uses + 1 WHERE id = ?");
        $stmt->execute([$coupon['id']]);

        $stmt = $db->prepare("SELECT * FROM orders WHERE id = ?");
        $stmt->execute([$id]);
        $order = $stmt->fetch();

        if (!$order) {
            $response->getBody()->write(json_encode([
                'error' => 'Not Found',
                'message' => 'Pedido nao encontrado',
            ]));
            return $response->withStatus(404)->withHeader('Content-Type', 'application/json');
        }

        $discount = round($order['total'] * ($coupon['discount'] / 100) * 100) / 100;
        $newTotal = round(($order['total'] - $discount) * 100) / 100;

        $stmt = $db->prepare("UPDATE orders SET total = ? WHERE id = ?");
        $stmt->execute([$newTotal, $id]);

        Database::logRequest([
            'method' => 'POST', 'path' => "/api/orders/$id/apply-coupon",
            'queryParams' => '', 'body' => json_encode($body),
            'statusCode' => 200,
            'sqlQuery' => "UPDATE coupons SET uses = uses + 1 WHERE id = {$coupon['id']}; UPDATE orders SET total = $newTotal WHERE id = $id",
            'responsePreview' => "coupon $code applied, discount $discount",
            'ip' => self::getIp($request),
        ]);

        $response->getBody()->write(json_encode([
            'message'        => 'Cupom aplicado com sucesso',
            'discount'       => $discount,
            'original_total' => $order['total'],
            'new_total'      => $newTotal,
            'coupon_code'    => $code,
            'coupon_uses'    => $coupon['uses'] + 1,
            'coupon_max'     => $coupon['max_uses'],
        ]));
        return $response->withHeader('Content-Type', 'application/json');
    }

    private static function getIp(Request $request): string
    {
        return $request->getHeaderLine('X-Forwarded-For') ?: ($request->getServerParams()['REMOTE_ADDR'] ?? 'unknown');
    }
}

File: 05-seguranca-ctf/backend/java/src/main/java/com/treinamento/ctf/controller/OrderController.java
package com.treinamento.ctf.controller;

import com.treinamento.ctf.middleware.RequestLogFilter;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.*;

import com.treinamento.ctf.config.DbUtil;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/orders")
public class OrderController {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @GetMapping
    public ResponseEntity<?> listOrders(HttpServletRequest request) {
        String userId = (String) request.getAttribute("userId");
        String role   = (String) request.getAttribute("userRole");

        String sql;
        List<Map<String, Object>> orders;

        if ("admin".equals(role)) {
            sql = "SELECT * FROM orders ORDER BY id DESC";
            RequestLogFilter.CURRENT_SQL.set(sql);
            orders = jdbcTemplate.queryForList(sql);
        } else {
            sql = "SELECT * FROM orders WHERE user_id = " + userId + " ORDER BY id DESC";
            RequestLogFilter.CURRENT_SQL.set(sql);
            orders = jdbcTemplate.queryForList(
                "SELECT * FROM orders WHERE user_id = ? ORDER BY id DESC", userId
            );
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("orders", DbUtil.lowercaseKeys(orders));
        result.put("count", orders.size());

        return ResponseEntity.ok(result);
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getOrderById(@PathVariable Integer id, HttpServletRequest request) {
        String sql = "SELECT o.*, u.nome as user_name, u.email as user_email " +
                     "FROM orders o JOIN users u ON o.user_id = u.id WHERE o.id = " + id;
        RequestLogFilter.CURRENT_SQL.set(sql);

        try {
            Map<String, Object> order = jdbcTemplate.queryForMap(
                "SELECT o.*, u.nome as user_name, u.email as user_email " +
                "FROM orders o JOIN users u ON o.user_id = u.id WHERE o.id = ?", id
            );

            return ResponseEntity.ok(DbUtil.lowercaseKeys(order));
        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Not Found");
            err.put("message", "Pedido nao encontrado");
            return ResponseEntity.status(404).body(err);
        }
    }

    @PostMapping
    public ResponseEntity<?> createOrder(@RequestBody Map<String, Object> body,
                                         HttpServletRequest request) {
        String userId = (String) request.getAttribute("userId");

        List<?> items = (List<?>) body.get("items");
        Number totalNum = (Number) body.get("total");
        double total = totalNum != null ? totalNum.doubleValue() : 0;

        if ((items == null || items.isEmpty()) || total <= 0) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Bad Request");
            err.put("message", "Pedido deve conter itens e total valido");
            return ResponseEntity.status(400).body(err);
        }

        String sqlQuery = "INSERT INTO orders (user_id, total, status) VALUES (" +
                userId + ", " + total + ", 'pendente')";
        RequestLogFilter.CURRENT_SQL.set(sqlQuery);

        try {
            jdbcTemplate.update(
                "INSERT INTO orders (user_id, total, status) VALUES (?, ?, 'pendente')",
                Integer.parseInt(userId), total
            );

            Map<String, Object> newOrder = jdbcTemplate.queryForMap(
                "SELECT * FROM orders WHERE id = (SELECT MAX(id) FROM orders)"
            );

            Map<String, Object> result = new LinkedHashMap<>();
            result.put("message", "Pedido criado com sucesso");
            result.put("order", DbUtil.lowercaseKeys(newOrder));

            return ResponseEntity.status(201).body(result);
        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Database error");
            err.put("message", ex.getMessage());
            return ResponseEntity.status(500).body(err);
        }
    }

    @PostMapping("/{id}/apply-coupon")
    public ResponseEntity<?> applyCoupon(@PathVariable Integer id,
                                         @RequestBody Map<String, String> body,
                                         HttpServletRequest request) {
        String code = body.getOrDefault("code", "");

        if (code.isEmpty()) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Bad Request");
            err.put("message", "Codigo do cupom obrigatorio");
            return ResponseEntity.status(400).body(err);
        }

        Map<String, Object> coupon;
        try {
            coupon = jdbcTemplate.queryForMap(
                "SELECT * FROM coupons WHERE code = ? AND active = 1", code
            );
        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Not Found");
            err.put("message", "Cupom nao encontrado");
            return ResponseEntity.status(404).body(err);
        }

        int uses = ((Number) coupon.get("USES")).intValue();
        int maxUses = ((Number) coupon.get("MAX_USES")).intValue();

        if (uses >= maxUses) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Bad Request");
            err.put("message", "Cupom esgotado");
            return ResponseEntity.status(400).body(err);
        }

        // Increment coupon usage
        Integer couponId = (Integer) coupon.get("ID");
        jdbcTemplate.update("UPDATE coupons SET uses = uses + 1 WHERE id = ?", couponId);

        Map<String, Object> order;
        try {
            order = jdbcTemplate.queryForMap("SELECT * FROM orders WHERE id = ?", id);
        } catch (Exception ex) {
            Map<String, Object> err = new LinkedHashMap<>();
            err.put("error", "Not Found");
            err.put("message", "Pedido nao encontrado");
            return ResponseEntity.status(404).body(err);
        }

        double orderTotal = ((Number) order.get("TOTAL")).doubleValue();
        double discountPct = ((Number) coupon.get("DISCOUNT")).doubleValue();
        double discount = Math.round(orderTotal * (discountPct / 100) * 100.0) / 100.0;
        double newTotal = Math.round((orderTotal - discount) * 100.0) / 100.0;

        jdbcTemplate.update("UPDATE orders SET total = ? WHERE id = ?", newTotal, id);

        RequestLogFilter.CURRENT_SQL.set(
            "UPDATE coupons SET uses = uses + 1 WHERE id = " + couponId +
            "; UPDATE orders SET total = " + newTotal + " WHERE id = " + id
        );

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("message", "Cupom aplicado com sucesso");
        result.put("discount", discount);
        result.put("original_total", orderTotal);
        result.put("new_total", newTotal);
        result.put("coupon_code", code);
        result.put("coupon_uses", uses + 1);
        result.put("coupon_max", maxUses);

        return ResponseEntity.ok(result);
    }
}
